### ISSUE TITLE
Why truffle npm package is distributed in a minified version?

### ISSUE THREAD

**Issue by chebykin** [trufflesuite/truffle#2440]
- [ no] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue.

---------------------------

## Issue

Hi! I'm curious why `truffle` npm package is distributed in an already minified version? I can imagine only the following cases:

- if I use it with front-end, it's a big chance that I do minification using my own `webpack` config or another tool
- if I use it with `node.js`, I probably don't care about it at all
- if I use it to develop Solidity contracts, I would probably like to make some hacks or debug original truffle code

But none of these cases requires minification in a first glance.

I'm just stumbled into it when trying to hack truffle code right in my `node_modules` folder.

**Comment by eggplantzzz** [comment]
Hey @chebykin! The main entry file for Truffle is actually not minified. Will you tell me why you think it is minified? Do you have something specific that you can refer me to?

**Comment by chebykin** [comment]
I'm sorry, you are right, it isn't minified. That was my mistake, I had incorrectly used this term.

The question should sound `Why almost all truffle modules are bundled into a single file?`. By the single file, I mean `node_modules/truffle/build/cli.bundled.js`.

The problem is that it's too difficult to edit a file that is over 500K lines of code. All of my code editors hang when I trying to do this 😀.

**Comment by eggplantzzz** [comment]
Ah ok, the bundled file is the binary that is executed. However, if you want to edit that file, you shouldn't do it directly. Instead you should make changes in the necessary packages and then rebuild that file.

If you need any guidance on what to edit and where then feel free to ask more questions! Another thing that might help a bit for gaining an understanding of the flow of Truffle, you can check out the [CONTRIBUTING.md](https://github.com/trufflesuite/truffle/blob/develop/CONTRIBUTING.md) which contains a little bit of general information about Truffle's architecture.

**Comment by gnidan** [comment]
To add: the bundle exists to provide a mechanism to ensure portability. In the past (and maybe today, even), a number of Truffle's dependencies have required native dependencies. Our solution has been to bundle into a single Node-only module.

**Comment by chebykin** [comment]
Ok, I see that things are more complex than I thought. Thank you for the answers!

If someone is looking for including non-bundled truffle packages I found the following hack. In your `package.json` file instead of including `truffle` package into your dependencies, include `@truffle/core` and `@truffle/environment`:

```javascript
{
    "@truffle/core": "^5.0.38",
    "@truffle/environment": "^0.1.14",
    ...
}
```

I was able to compile my contracts and run test cases, but there was a problem with BN object compatibility which doesn't appear in the full bundled `truffle` version. I didn't check migrations too.

I've got my answer😊 so I'm closing the ticket. Thanks again!
